bootstore/src/schemes/v0/state_initial_member.rs

```diff
@@ -0,0 +1,411 @@
+
+    // Pending learn requests from other peers mapped to their start time.
+    //
+    // When a peer attempts to learn a share, we may not be able to give it
+    // one because we cannot yet recompute the rack secret. We queue requests
+    // here and respond when we can recompute the rack secret. If we timeout before
+    // we can recompute the rack secret, we respond with a timeout Error.
+    //
+    // Note that if we get a new `RequestType::Learn` from a peer that is already
+    // pending, we will reset the start time.
+    pub pending_learn_requests: BTreeMap<Baseboard, RequestMetadata>,
+}
+
+impl InitialMemberState {
+    pub fn new(
+        pkg: SharePkgV0,
+        distributed_shares: BTreeMap<Baseboard, ShareIdx>,
+    ) -> Self {
+        InitialMemberState {
+            pkg,
+            distributed_shares,
+            rack_init_state: None,
+            pending_learn_requests: BTreeMap::new(),
+        }
+    }
+
+    pub fn name(&self) -> &'static str {
+        "initial_member"
+    }
+
+    // Process a `Share` received in a `Response`
+    pub fn on_share(
+        &mut self,
+        common: &mut FsmCommonData,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Vec<u8>,
+    ) -> Output {
+        // The RackSecret state needs this deadline to know how long to keep
+        // the secret
+        // TODO: Use a separate timeout for this purpose?
+        let rack_secret_expiry =
+            common.clock + common.config.rack_secret_request_timeout;
+
+        let mut output = common.rack_secret_state.on_share(
+            from,
+            request_id,
+            share,
+            rack_secret_expiry,
+        );
+
+        if let Some(Ok(ApiOutput::RackSecret(rack_secret))) =
+            &mut output.api_output
+        {
+            // We have the rack secret as part of `output`. We may also
+            // have learn request that we need to resolve now.
+            //
+            // Note that we only persist if there are learn requests which
+            // cause an update to persistent state by handing out a share to the requester.
+            let (persist, envelopes) =
+                self.resolve_learn_requests(&rack_secret);
+            output.envelopes.extend_from_slice(&envelopes);
+            output.persist = persist;
+        }
+        output
+    }
+
+    // We've just recomputed the rack secret. Now resolve any pending learn
+    // requests. Return whether we need to persist state and any envelopes to
+    // send to peers.
+    fn resolve_learn_requests(
+        &mut self,
+        rack_secret: &RackSecret,
+    ) -> (bool, Vec<Envelope>) {
+        let mut persist = false;
+        let mut envelopes =
+            Vec::with_capacity(self.pending_learn_requests.len());
+        match self.pkg.decrypt_shares(rack_secret) {
+            Ok(shares) => {
+                while let Some((to, metadata)) =
+                    self.pending_learn_requests.pop_first()
```

Comment:
No need to peel off items once by one:
```suggestion
                for (to, metadata) in
                    std::mem::take(&mut self.pending_learn_requests)
```
(this also lets you remove the clone in `to.clone()` below)

Reply:
Interesting. I may have used this pattern elsewhere. I'll keep this comment open and go through the code.